kratos/utilities/parallel_utilities.h

```diff
@@ -38,6 +38,61 @@ namespace Kratos
 {
 ///@addtogroup KratosCore
 
+/// Shared memory parallelism related helper class
+/** Provides access to functionalities for shared memory parallelism
+ * such as the number of threads in usa.
+*/
+class KRATOS_API(KRATOS_CORE) ParallelUtilities
+{
+public:
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    ParallelUtilities() = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /** @brief Returns the current number of threads
+     * @return number of threads
+     */
+    static int GetNumThreads();
+
+    /** @brief Sets the current number of threads
+     * @param NumThreads - the number of threads to be used
+     */
+    static void SetNumThreads(const int NumThreads);
+
+    /** @brief Returns the number of processors available to this device
+     * This can include the multiple threads per processing unit
+     * @return number of processors
+     */
+    static int GetNumProcs();
+
+    ///@}
+
+private:
+    ///@name Static Member Variables
+    ///@{
+
+    static int msNumThreads; // number of threads used in Kratos
```

Comment:
I would suggest using at least Meyers' singleton instead of having a static member. 

Using the static member may cause the know static initialization order fiasco problem which may lead to random failures difficult to find!

Reply:
ok I was not sure if this is a must, thx for clarifying

Reply:
I implemented the singleton, can you please have a look @pooyan-dadvand ?

Reply:
I am not sure having the double guard here would affect the performance in the usage or not. If it affects, I would go for the simpler Meyers' singleton which I think would work fine here.

Reply:
Then I am lost :/
I googled and checked what we have so far in Kratos (=> parallel_environment.h)
I implemented basically what (according to my research) is the simplest version of the meyer-singleton

can you point me to another reference?

Reply:
I think I figured it out, can you please have a look?